src/box/txn_limbo.c

```diff
@@ -586,7 +586,8 @@ txn_limbo_ack(struct txn_limbo *limbo, uint32_t replica_id, int64_t lsn)
 {
 	if (rlist_empty(&limbo->queue))
 		return;
-	if (limbo->frozen)
+	/** Do not ack anything while being read-only. */
+	if (limbo->frozen || box_is_ro())
```

Comment:
> That looks dangerous. What if I set box.cfg.read_only to true after a transaction is started and then collect ACKs? No new leader and I could write a CONFIRM just fine, but here I won't. Also, when if I set it back to false afterwards, the transaction would be stuck and rolled back on timeout?

Good point. As discussed in the chat, I've reworked it so that limbo now has 2 reasons for being frozen:
1. fencing
2. recovering a limbo owner and waiting for at least one PROMOTE to happen

Bullet 2 covers the case I was trying to fix with `box_is_ro()`.

> Besides, the issue can happen even after recovery, if all local data is restored, but replication is still catching up with the other nodes. Isn't it true?

Agree.

> The behaviour of collecting ACKs after restart looks legitimate, I think its absence was even a bug at some point.

We agreed it was a bug in a private chat. The reasoning is in commit message for commit "txn_limbo: do not confirm/rollback anything after restart"

> Or do you mean CONFIRMs are being written right during recovery even if there are CONFIRMs written later in WAL? Then we could add a new function txn_limbo_recover(request), like raft_process_recovery(request) which solves a similar problem.

No. You understood me correctly from the beginning. I meant that the node recovers not-applied txns, and then starts collecting ACKs even before checking if anyone else has become the new leader